internal/pubsub/rabbitpubsub/rabbit.go

```diff
@@ -180,3 +183,164 @@ func (*topic) IsRetryable(error) bool {
 	return false
 }
+
+// OpenSubscription returns a *pubsub.Subscription corresponding to the named queue. The
+// queue must have been previously created (for instance, by using
+// amqp.Channel.QueueDeclare) and bound to an exchange.
+//
+// OpenSubscription uses the supplied amqp.Connection for all communication. It is
+// the caller's responsibility to establish this connection before calling OpenSubscription
+// and to close it when Close has been called on all Subscriptions opened with it.
+//
+// The documentation of the amqp package recommends using separate connections for
+// publishing and subscribing.
+func OpenSubscription(conn *amqp.Connection, name string) (*pubsub.Subscription, error) {
+	// TODO(jba): support context.Context
+	s, err := newSubscription(conn, name)
+	if err != nil {
+		return nil, err
+	}
+	return pubsub.NewSubscription(s), nil
+}
+
+type subscription struct {
+	conn     *amqp.Connection
+	queue    string // the AMQP queue name
+	consumer string // the client-generated name for this particular subscriber
+
+	mu   sync.Mutex
+	ch   *amqp.Channel // AMQP channel used for all communication.
+	delc <-chan amqp.Delivery
+}
+
+var nextConsumer int64 // atomic
+
+func newSubscription(conn *amqp.Connection, name string) (*subscription, error) {
+	s := &subscription{
+		conn:     conn,
+		queue:    name,
+		consumer: fmt.Sprintf("c%d", atomic.AddInt64(&nextConsumer, 1)),
+	}
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if err := s.establishChannel(); err != nil {
+		return nil, err
+	}
+	return s, nil
+}
+
+// Must be called with s.mu held.
+func (s *subscription) establishChannel() error {
+	// TODO(jba): support context.Context
+	if s.ch != nil {
+		// We already have a channel; nothing to do.
+		return nil
+	}
+	// Create a new channel.
+	ch, err := s.conn.Channel()
+	if err != nil {
+		return err
+	}
+	// Subscribe to messages from the queue.
+	s.delc, err = ch.Consume(s.queue, s.consumer,
+		false, // autoAck
+		false, // exclusive
+		false, // noLocal
+		wait,
+		nil) // args
+	if err != nil {
+		return err
+	}
+	s.ch = ch
+	return nil
+}
+
+// ReceiveBatch implements driver.Subscription.ReceiveBatch.
+func (s *subscription) ReceiveBatch(ctx context.Context, maxMessages int) ([]*driver.Message, error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if err := s.establishChannel(); err != nil {
+		return nil, err
+	}
+
+	// Get up to maxMessages waiting messages, but don't take too long.
+	var ms []*driver.Message
+	maxTime := time.After(50 * time.Millisecond)
+	for {
+		select {
+		case <-ctx.Done():
+			// Cancel the Consume.
+			_ = s.ch.Cancel(s.consumer, wait) // ignore the error
+			s.ch = nil
+			return nil, ctx.Err()
+
+		case d, ok := <-s.delc:
+			if !ok { // channel closed
+				s.ch = nil // re-establish the channel next time
+				if len(ms) > 0 {
+					return ms, nil
+				}
+				// TODO(jba): get more information for a better message
+				return nil, errors.New("delivery channel closed")
```

Comment:
This is retryable, right?

Reply:
I need to understand better why it would be closed. That's one of my next tasks.